File: src/progress.ts
/**
 * Terminal progress reporting for megadj: one line per item plus a live
 * bar with rate + ETA when stdout is a TTY, plain milestone lines otherwise.
 */

const BAR_WIDTH = 24;
// stderr, always: stdout belongs to the data (--json must stay parseable,
// piped output must stay clean). TTY detection follows stderr accordingly.
const isTty = process.stderr.isTTY ?? false;

export function fmtBytes(n: number): string {
  const units = ["B", "KB", "MB", "GB", "TB"];
  let v = n;
  let u = 0;
  while (v >= 1024 && u < units.length - 1) {
    v /= 1024;
    u++;
  }
  return u === 0 ? `${Math.round(v)} B` : `${v.toFixed(1)} ${units[u]}`;
}

export function fmtDur(seconds: number): string {
  const s = Math.max(0, Math.round(seconds));
  const h = Math.floor(s / 3600);
  const m = Math.floor((s % 3600) / 60);
  const sec = s % 60;
  return h > 0
    ? `${h}:${String(m).padStart(2, "0")}:${String(sec).padStart(2, "0")}`
    : `${String(m).padStart(2, "0")}:${String(sec).padStart(2, "0")}`;
}

/** Live single-line progress bar with rate and ETA. */
export class ProgressBar {
  private readonly start = Date.now();
  private done = 0;
  private bytes = 0;
  private lastRender = 0;
  private lastPct = -1;

  constructor(
    private readonly total: number,
    private readonly label: string,
    private readonly unit = "tracks",
  ) {}

  update(n = 1, bytes = 0): void {
    this.done += n;
    this.bytes += bytes;
    this.render();
  }

  private render(force = false): void {
    const now = Date.now();
    if (!force && now - this.lastRender < 250) return;
    this.lastRender = now;
    const elapsed = (now - this.start) / 1000;
    const pct = this.total > 0 ? this.done / this.total : 0;

    if (!isTty) {
      const pctInt = Math.floor(pct * 100);
      if (pctInt >= this.lastPct + 5 || force) {
        this.lastPct = pctInt;
        process.stderr.write(
          `[${this.label}] ${this.done}/${this.total} ${this.unit} · ${pctInt}% · ${this.suffix(elapsed)}\n`,
        );
      }
      return;
    }

    const filled = Math.round(BAR_WIDTH * Math.min(pct, 1));
    const bar = "#".repeat(filled) + "-".repeat(BAR_WIDTH - filled);
    const line = `[${this.label}] [${bar}] ${this.done}/${this.total} ${this.unit} · ${Math.min(Math.round(pct * 100), 100)}% · ${this.suffix(elapsed)}`;
    process.stderr.write(`\r\u001b[K${line.slice(0, 120)}`);
  }

  private suffix(elapsed: number): string {
    const parts: string[] = [];
    if (this.bytes > 0 && elapsed > 1)
      parts.push(`${fmtBytes(this.bytes / elapsed)}/s`);
    if (this.done > 0 && elapsed > 2) {
      const remaining = (elapsed / this.done) * (this.total - this.done);
      parts.push(`ETA ${fmtDur(remaining)}`);
    }
    return parts.join(" · ");
  }

  close(summary?: string): void {
    const elapsed = (Date.now() - this.start) / 1000;
    this.render(true);
    if (isTty) process.stderr.write("\n");
    const parts = [
      `[${this.label}] ${this.done}/${this.total} ${this.unit} in ${fmtDur(elapsed)}`,
    ];
    if (this.bytes > 0)
      parts.push(
        `(${fmtBytes(this.bytes)} at ${fmtBytes(this.bytes / Math.max(elapsed, 0.001))}/s)`,
      );
    process.stderr.write(`${parts.join(" ")}\n`);
    if (summary) console.log(summary); // explicit caller-provided summary: stdout
  }
}

/** The standard command log function (P1 --json contract): human progress
 * goes to stdout via console.log normally; in --json mode stdout carries
 * exactly one summary object, so the human log moves to STDERR (never
 * silent — the CrateDeck intake job streams these lines as its live phase
 * progress, and a swallowed log made a run look frozen at 0%).
 * `onProgress` still owns routing when the caller injected one. Every
 * command's `const log = ...` preamble was this exact three-liner — now
 * one call. */
export function commandLog(opts: {
  json?: boolean | undefined;
  onProgress?: ((msg: string) => void) | undefined;
}): (msg: string) => void {
  if (opts.onProgress) return opts.onProgress;
  if (opts.json) return (m) => process.stderr.write(`${m}\n`);
  return (m: string) => console.log(m);
}

File: plugin/skills/rekordbox-usb-sync/scripts/progress.py
#!/usr/bin/env python3
"""Terminal progress bar with ETA + rate for the rekordbox-usb-sync scripts.

Zero dependencies. Renders a single updating line on a TTY:

    Contents copy | [########--------] 1450/3794 files · 1.2/4.5 GB · 27% · 4.2 MB/s · ETA 03:12

When stdout is not a TTY (piped to a log), prints one plain line every 5%
instead so `tail -f` still shows movement.
"""

import sys
import time


def fmt_bytes(n: float) -> str:
    for unit in ("B", "KB", "MB", "GB", "TB"):
        if n < 1024 or unit == "TB":
            return f"{int(n)} B" if unit == "B" else f"{n:.1f} {unit}"
        n /= 1024
    return f"{n:.1f} TB"


def fmt_dur(seconds: float) -> str:
    s = int(seconds)
    h, rem = divmod(s, 3600)
    m, sec = divmod(rem, 60)
    if h:
        return f"{h}:{m:02d}:{sec:02d}"
    return f"{m:02d}:{sec:02d}"


class Progress:
    """Track items (+ optional bytes) and render a live bar with ETA."""

    def __init__(
        self,
        total: int,
        label: str = "",
        unit: str = "files",
        total_bytes: int | None = None,
        initial: int = 0,
    ):
        self.total = max(total, 1)
        self.label = label
        self.unit = unit
        self.total_bytes = total_bytes
        self.done = initial
        self.bytes_done = 0
        self.start = time.monotonic()
        self._last_render = 0.0
        self._last_plain = -1
        self.is_tty = sys.stdout.isatty()

    def update(self, n: int = 1, nbytes: int = 0) -> None:
        self.done += n
        self.bytes_done += nbytes
        self.render()

    def render(self, force: bool = False) -> None:
        now = time.monotonic()
        if not force and now - self._last_render < 0.25:
            return
        self._last_render = now
        elapsed = now - self.start
        if not self.is_tty:
            pct = self.done * 100 // self.total
            if pct >= self._last_plain + 5 or force:
                self._last_plain = pct
                print(
                    f"[{self.label}] {self.done}/{self.total} {self.unit} · {self._suffix(elapsed)}",
                    flush=True,
                )
            return
        width = 24
        filled = int(width * min(self.done / self.total, 1.0))
        bar = "#" * filled + "-" * (width - filled)
        line = f"[{bar}] {self.done}/{self.total} {self.unit} · {self._suffix(elapsed)}"
        if self.label:
            line = f"{self.label} | {line}"
        sys.stdout.write("\r\033[K" + line[:120])
        sys.stdout.flush()

    def _suffix(self, elapsed: float) -> str:
        parts = [f"{min(self.done * 100 // self.total, 100):3d}%"]
        if self.total_bytes:
            parts.append(
                f"{fmt_bytes(self.bytes_done)} / {fmt_bytes(self.total_bytes)}"
            )
            if elapsed > 1 and self.bytes_done > 0:
                parts.append(f"{fmt_bytes(self.bytes_done / elapsed)}/s")
        if elapsed > 2 and self.done > 0:
            remaining = (elapsed / self.done) * (self.total - self.done)
            parts.append("ETA " + fmt_dur(remaining))
        return " · ".join(parts)

    def close(self, message: str | None = None) -> None:
        elapsed = time.monotonic() - self.start
        self.render(force=True)
        if self.is_tty:
            sys.stdout.write("\n")
            sys.stdout.flush()
        summary = f"[{self.label or 'done'}] {self.done}/{self.total} {self.unit} in {fmt_dur(elapsed)}"
        if self.total_bytes and elapsed > 0:
            summary += f" ({fmt_bytes(self.bytes_done)} at {fmt_bytes(self.bytes_done / elapsed)}/s)"
        print(summary, flush=True)
        if message:
            print(message, flush=True)


class Stage:
    """Named stage banners so multi-step runs read like a job log."""

    def __init__(self) -> None:
        self.n = 0
        self.start = time.monotonic()

    def step(self, name: str) -> None:
        self.n += 1
        print(
            f"\n=== [{self.n}] {name}  (t+{fmt_dur(time.monotonic() - self.start)}) ===",
            flush=True,
        )

    def info(self, msg: str) -> None:
        print(f"    {msg}", flush=True)
